data/src/user.rs

```diff
@@ -18,39 +19,190 @@ pub struct User {
     username: Option<String>,
     hostname: Option<String>,
     accountname: Option<String>,
-    access_levels: HashSet<AccessLevel>,
+    access_levels: BTreeSet<AccessLevel>,
     away: bool,
 }
 
+impl Hash for User {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.nickname.hash(state);
+    }
+}
+
+impl Ord for User {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.access()
+            .cmp(&other.access())
+            .reverse()
+            .then_with(|| self.nick().cmp(&other.nick()))
+    }
+}
+impl PartialOrd for User {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
 impl PartialEq for User {
     fn eq(&self, other: &Self) -> bool {
-        self.nickname.eq(&other.nickname)
+        self.nick() == other.nick() && self.access() == other.access()
     }
 }
-
 impl Eq for User {}
 
-impl Hash for User {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        self.nickname.hash(state);
+#[derive(Debug, Default)]
+pub struct ChannelUsers(BTreeSet<User>);
```

Comment:
I'm worried this can introduce a sneaky error where we might accidentally add the same user twice because we have different access levels at different times.

We can use the same approach you have in `get_by_nick` of checking against each access level for the added / removed nickname to safeguard this. Considering its a binary search, its not too much extra cost.

Alternatively, we scrap this entirely as its a bit of complexity now, and we still do something akin to "sync" as before, but we only call "sync" after certain commands come in that actually mutate the channel / user state.

Reply:
Yes, this has been on my mind too.

I think we should keep this design for two reasons:
1. I think that odd bugs could sneak into the other implementation too. The previous implementation of PartialEq and Ord disagree, which is not allowed and could cause issues.
2. I think the set API is useful for downstream files, and they can offer much more efficient operations with it.


I think a better idea might be to get rid of all of the magic, and keep the previous `Eq` and `Hash` implementations. Then, we could let `ChannelUsers` be an `IndexMap`, which will always match when indexed by a `User`, but it is ordered correctly.

Then care is taken to insert the user to the right location in the channel on `insert` (which is basically what you are suggesting but without a clone).